db/dcrpg/pgblockchain.go

```diff
@@ -571,37 +626,91 @@ func (pgb *ChainDB) GetTicketPoolByDateAndInterval(maturityBlock int64,
 	return retrieveTicketsByDate(pgb.db, maturityBlock, int64(val))
 }
 
-// TicketPoolVisualization fetches the following ticketpool data: tickets
+// TicketPoolVisualization helps block consecutive access to the ticketPoolVisualization
+// method till the first request is resolved.
+func (pgb *ChainDB) TicketPoolVisualization(interval dbtypes.ChartGrouping) ([]*dbtypes.PoolTicketsData, *dbtypes.PoolTicketsData, uint64, error) {
+	// Attempt to retrieve data for the current block from cache.
+	heightSeen := pgb.Height()
+	barcharts, donutCharts, height, ok := TicketPoolData(interval, heightSeen)
+	if ok {
+		// The cache was fresh.
+		return barcharts, donutCharts, height, nil
+	}
+
+	// Cache is stale. Attempt to gain updater status.
+	if !pgb.tpUpdatePermission.TryLock() {
+		// Another goroutine is running db query to get the updated data. Return
+		// stale data instead of waiting.
+		return barcharts, donutCharts, height, nil
+	}
+	// This goroutine is now the cache updater.
+	defer pgb.tpUpdatePermission.Unlock()
+
+	// Retrieve chart data for best block in DB.
+	var err error
+	barcharts, donutCharts, height, err = pgb.ticketPoolVisualization(interval)
+	if err != nil {
+		log.Errorf("Failed to fetch ticket pool data: %v", err)
+		return nil, nil, 0, err
+	}
+
+	// Update the cache with the new ticket pool data.
+	UpdateTicketPoolData(interval, barcharts, donutCharts, height)
+
+	return barcharts, donutCharts, height, nil
+}
+
+// ticketPoolVisualization fetches the following ticketpool data: tickets
 // grouped on the specified interval, tickets grouped by price, and ticket
 // counts by ticket type (solo, pool, other split). The interval may be one of:
 // "mo", "wk", "day", or "all". The data is needed to populate the ticketpool
 // graphs. The data grouped by time and price are returned in a slice.
-func (pgb *ChainDB) TicketPoolVisualization(interval dbtypes.ChartGrouping) (
-	[]*dbtypes.PoolTicketsData, *dbtypes.PoolTicketsData, error) {
-	var maturityBlock = pgb.GetTicketPoolBlockMaturity()
-
-	// Tickets grouped by time interval
-	ticketsByTime, err := pgb.GetTicketPoolByDateAndInterval(maturityBlock, interval)
+func (pgb *ChainDB) ticketPoolVisualization(interval dbtypes.ChartGrouping) (byTimeAndPrice []*dbtypes.PoolTicketsData, byInputs *dbtypes.PoolTicketsData, height uint64, err error) {
+	// Ensure DB height is the same before and after queries since they are not
+	// atomic. Initial height:
+	height, err = pgb.HeightDB()
 	if err != nil {
-		return nil, nil, err
+		err = fmt.Errorf("failed to query DB height: %v", err)
+		return
 	}
 
-	// Tickets grouped by price
-	ticketsByPrice, err := retrieveTicketByPrice(pgb.db, maturityBlock)
-	if err != nil {
-		return nil, nil, err
-	}
+	for {
```

Comment:
I don't get why we need a for loop here, the queries will obviously return the latest data as in the db. This may also go against our earlier priority of serving first and fast just incase block heights keeps changing in split seconds.

Reply:
There are multiple queries.  They all need to be done when the DB is at the same height

Reply:
It's not likely to loop, but this guarantees consistency

Reply:
ok. Noted